crates/wasmtime/src/runtime/component/func/options.rs

```diff
@@ -497,19 +502,40 @@ impl<'a> LiftContext<'a> {
         let memory = options
             .memory
             .map(|_| options.memory(unsafe { &*(store as *const StoreOpaque) }));
-        let (calls, host_table, host_resource_data, instance) =
-            store.component_resource_state_with_instance(instance_handle);
-        let (component, instance) = instance.component_and_self();
-
-        LiftContext {
-            memory,
-            options,
-            types: component.types(),
-            instance,
-            instance_handle,
-            calls,
-            host_table,
-            host_resource_data,
+        #[cfg(feature = "component-model-async")]
+        {
+            let (calls, host_table, host_resource_data, instance, concurrent_state) =
+                store.component_resource_state_with_instance_and_concurrent_state(instance_handle);
+            let (component, instance) = instance.component_and_self();
+
+            LiftContext {
+                memory,
+                options,
+                types: component.types(),
+                instance,
+                instance_handle,
+                calls,
+                host_table,
+                host_resource_data,
+                concurrent_state,
+            }
+        }
+        #[cfg(not(feature = "component-model-async"))]
+        {
+            let (calls, host_table, host_resource_data, instance) =
+                store.component_resource_state_with_instance(instance_handle);
+            let (component, instance) = instance.component_and_self();
+
+            LiftContext {
+                memory,
+                options,
+                types: component.types(),
+                instance,
+                instance_handle,
+                calls,
+                host_table,
+                host_resource_data,
+            }
         }
```

Comment:
Would it makes sense to have a `concurrent_disabled.rs` which has ZST for `ConcurrentState` with no methods? That would avoid this `#[cfg]` for example and help encapsulate "all the conditional stuff is in this file" too